include/zephyr/drivers/crc.h

```diff
@@ -0,0 +1,191 @@
+/*
+ * Copyright (c) 2024 Brill Power Ltd.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_INCLUDE_DRIVERS_CRC_H
+#define ZEPHYR_INCLUDE_DRIVERS_CRC_H
+
+#include <zephyr/device.h>
+#include <zephyr/kernel.h>
+
+#ifdef __cplusplus
+extern "C" {
+#endif
+
+/**
+ * @brief CRC driver APIs
+ * @defgroup crc_interface CRC driver APIs
+ * @ingroup io_interfaces
+ * @{
+ */
+
+#define CRC_FLAG_REVERSE_INPUT  BIT(0)
+#define CRC_FLAG_REVERSE_OUTPUT BIT(1)
+
+#define CRC4_INIT_VAL        0x0
+#define CRC4_TI_INIT_VAL     0x0
+#define CRC7_BE_INIT_VAL     0x0
+#define CRC8_INIT_VAL        0x0
+#define CRC8_CCITT_INIT_VAL  0x00
+#define CRC16_INIT_VAL       0x0
+#define CRC16_ANSI_INIT_VAL  0x0
+#define CRC16_CCITT_INIT_VAL 0x0000
+#define CRC16_ITU_T_INIT_VAL 0x0000
+#define CRC24_PGP_INIT_VAL   0x0
+#define CRC32_C_INIT_VAL     0x00000000
+#define CRC32_IEEE_INIT_VAL  0xFFFFFFFFU
+
+/**
+ * @brief CRC algorithm enumeration
+ */
+enum crc_type {
+	CRC4,
+	CRC4_TI,
+	CRC7_BE,
+	CRC8,
+	CRC8_CCITT,
+	CRC16,
+	CRC16_ANSI,
+	CRC16_CCITT,
+	CRC16_ITU_T,
+	CRC24_PGP,
+	CRC32_C,
+	CRC32_IEEE,
+};
+
+/**
+ * @brief CRC state enumeration
+ */
+enum crc_state {
+	CRC_STATE_IDLE,
+	CRC_STATE_IN_PROGRESS,
+	CRC_STATE_COMPLETE
+};
+
+/**
+ * @brief Runtime driver configuration
+ * @param mutex Zephyr mutex object to make driver thread-safe when driver
+ *		is used within multiple threads
+ */
+struct crc_data {
+	struct k_mutex mutex;
```

Comment:
Why have this for all drivers? This should be left up to the driver if concurrent calls need to be safeguarded IMO.